core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -83,18 +91,30 @@ protected HttpResponse onSuccess(ServiceRequestContext ctx, HttpRequest req) thr
      * Invoked when {@code req} is failed. By default, this method responds with the
      * {@link HttpStatus#UNAUTHORIZED} status.
      */
-    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req, @Nullable Throwable t)
+            throws Exception {
+        logger.warn("Unexpected exception during authorization.", t);
         final DefaultHttpResponse res = new DefaultHttpResponse();
         res.respond(HttpStatus.UNAUTHORIZED);
         return res;
     }
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        if (authorize(req.headers())) {
-            return onSuccess(ctx, req);
-        } else {
-            return onFailure(ctx, req);
-        }
+        DeferredHttpResponse res = new DeferredHttpResponse();
+
+        authorize(req, ctx).whenCompleteAsync((result, t) -> {
```

Comment:
```onSuccess```, the delegate service, needs to be run on the armeria event loop I think. If we don't use an async callback, we would probably run on, e.g. the Firebase auth client threadpool when using its client library.

Sync callbacks are appropriate when the callback logic is simple (like my previous ```.stream().anyMatch```) and especially where it doesn't make any additional async calls - then it doesn't really matter what thread runs the callback. But for callbacks that then run more async code like this one, we should make sure we're on the right thread, so I use ```whenCompleteAsync``` with the context's event loop.

With my new running authorizers in order, I use ```thenComposeAsync``` for similar reason, that ```Authorizer#authorize``` should be run on the event loop (compared to ```onSuccess```, I think there are less cases where it would be an actual problem not to though).